go/pkg/cas/upload.go

```diff
@@ -123,11 +177,13 @@ type UploadResult struct {
 
 // Digest returns the digest computed for a file/dir at ps.Path.
 //
-// To retrieve a digest of a regular file, only ps.Path is checked - other
-// fields are ignored.
-//
-// To retrieve a digest of a directory or a symlink, ps.Exclude must match one
+// If ps is a PathSpec pointer that was passed to Upload() earlier, then Digest
+// returns the digest computed for that PathSpec.
+// Otherwise PathSpec is matched structurally:
+//  - for regular files, only ps.Path is checked - other fields are ignored.
+//  - for directories and symlinks, ps.Exclude must match one
```

Comment:
Did you consider some sort of Context struct where the request (PathSpec) and result can live? That way you avoid documenting constraints around using the same pointer because its implicit. For instance:
```
type UploadContext struct {
  ps *PathSpec
  ur *UploadResult
}

func NewUploadContext(ps *PathSpec) *UploadContext {
}

func (c *UploadContext) RootDigest() *digest.Digest {
}

func (c *UploadContext) SubpathDigest(path string) *digest.Digest {
}
```

Reply:
I did not consider this. If we want to be more explicit and make APIs say what English says, then perhaps we could have

```go
// checks only pointer
func (r *UploadResult) Digest(*PathSpec) (digest.Digest, error)
func (r *UploadResult) DigestRegular(fileName string) (digest.Digest, error)
func (r *UploadResult) DigestDir(path string, exclude) (digest.Digest, error)
```

`DigestDir` isn't future proof though because if we add another PathSpec field that controls file traversal and must be propagated here, then we'd have to make a breaking change in DigestDir. The reason I proposed PathSpec struct (and called it "spec") is exactly this scenario: we can add new fields to PathSpec and they become the API of `UploadResult.Digest()`.

`DigestRegular` is future-proof though, since there is no file traversal. I can add that. WDYT?

Reply:
I did not fully understand the `UploadContext` idea or how it would in this case. What part of the contract-in-English does it alleviate? We still have to explain that Allowlist is ignored, and that Exclude must be same, and that none of them matter for regular files.

Reply:
The purpose of `UploadContext` is to avoid documenting that you need to pass the `PathSpec` with the exact same pointer to get the correct digest, which I feel is a bit awkward. Specially that PathSpec doesn't have any state, its a spec, and the requirement that it be the same pointer is an implementation detail of caching.  If the UploadResult is really `PathSpec` specific then imo we should tie these together so that either the `UploadResult` is retrievable from the `PathSpec` object or both are encapsulated in something like an `UploadContext`.

Reply:
OK, how about this

1) Remove UploadResult.Digest. Keep UploadResult though, as we might add something else later.
2) Rename PathSpec back to UploadItem. It was called PathSpec to be used with UploadResult.Digest(), but looks like we aren't happy with it
3) Add `UploadItem.Digest(relPath string) digest.Digest` which returns the digest for the path relative to `UploadItem.Path`. For example, `myItem.Digest(".")` returns the path for the root. It is pretty easy to use. One nice property is that it will work even for non-root ancestors for allowlisted-dirs.

This somewhat similar to the original proposal, where UploadItem.Digest is a callback, but this is simpler.

*For now*, Digest() will be guaranteed to work only after Upload() returns, but later can add `Done() <- chan struct{}` which is closed when Digest() is ready to be used.

WDYT?

Reply:
@rubensf

Reply:
Sgtm, I like point 3, that seems fairly simple to use.

Reply:
I'm not a big fan of a channel that indicates that data is available _somewhere else_. I don't think it's wrong but I'd hope it could be simpler than that. I _think_ it will look simpler to have a callback for results (if the user cares about them) - so instead of calling it `Digest()`, make it `OnComplete(ur UploadResults)` (as per other comment regarding channel inputs).

We could then add a `UploadResult.Digest(relPath string) digest.Digest`, instead of at the UploadItem. 

wdyt?

Reply:
> I'm not a big fan of a channel that indicates that data is available somewhere else. 

Isn't this the standard practice in Go? https://pkg.go.dev/context#Context.Done
A `chan struct{}` closed to indicate an event seems pretty standard: https://blog.golang.org/pipelines

> OnComplete(UploadResults)

is this UploadResults different from the one we have today? Or are you proposing to add a new struct? A func returning a channel seems simpler.

This, seems more complicated to use. The digests are guaranteed to be computed after Upload() succeeds. So if we ensured that it succeeded, then we can just call `Digest`. Compare:

```go
// simple
dig := in.Digest(".")

// complex
var dig digest.Digest
input.OnCompleted(func(r UploadResults) { 
  dig = r.Digest(".")
})
```

Reply:
also in general, once cannot `select` on a callback. How would a separate goroutine "block" on `OnComplete`?

also this would be more complicated to implement, since we'd need to store the callbacks, and then call them all, etc.

Anyway, let's continue in https://github.com/bazelbuild/remote-apis-sdks/pull/363
I didn't see your comment when I uploaded the new PR.

Reply:
Apologies, I had a different idea in mind of what you meant with the `Done` function. #363 lgtm.

Reply:
done